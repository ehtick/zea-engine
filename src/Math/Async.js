import { Signal } from '../Math/Signal';

class Async {
    constructor() {
        this.__asyncCount = 0;
        this.ready = new Signal(true);

        this.incAsyncCount = function(count=1) {
            this.__asyncCount+=count;
        }.bind(this);

        this.decAsyncCount = function() {
            if (this.__asyncCount > 0){
                this.__asyncCount--;
                if (this.__asyncCount == 0) {
                    this.__asyncsCompleted();
                }
            }
        }.bind(this);

        this.__asyncsCompleted = function(){
            this.ready.emit();
        }.bind(this)
    }

    get count(){
        return this.__asyncCount;
    }

};

export {
    Async
};
// export default Async;

