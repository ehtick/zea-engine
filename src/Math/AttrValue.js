// eslint-disable-next-line camelcase
import { JSON_stringify_fixedPrecision } from './Common.js'

/**
 * Base class for values that can be stored in vertex attributes.
 * <br>
 * **Note:** These values use {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Float32Array|Float32Array} values to store their data.
 */
class AttrValue {
  /**
   * Verifies that all the numeric values inside the `this.__data` variable are a valid number.
   * Returns `false` If at least one of the values is either {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Referencia/Objetos_globales/Infinity|Infinity} or
   * {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Referencia/Objetos_globales/NaN|NaN}.
   *
   * @return {boolean} - Returns the result as a boolean.
   */
  isValid() {
    for (const v of this.__data) {
      if (v == Infinity || isNaN(v)) return false
    }

    return true
  }

  /**
   * Creates a new value to wrap memory in an existing buffer.
   *
   * @param {ArrayBuffer} buffer - the buffer value.
   * @param {number} offset - the offset value.
   *
   * @todo Implement this function
   */
  static createFromFloat32Buffer(buffer, offset) {
    throw new Error(
      'Not yet implemented for this type:' + this.constructor.name
    )
  }

  /**
   * Returns the number of Float32 elements used by this type. Used to calculate storage requirements for large arrays of this type.
   *
   * @todo Implement this function
   */
  static numElements() {
    throw new Error(
      'Not yet implemented for this type:' + this.constructor.name
    )
  }

  /**
   * Returns current Vec2 data as array. Often used to pass types to the GPU.
   *
   * @return {array} - Returns the result as an array.
   */
  asArray() {
    return this.__data
  }

  /**
   * Calls `toJSON` method and stringifies it.
   *
   * @return {string} - The return value.
   */
  toString() {
    // eslint-disable-next-line new-cap
    return JSON_stringify_fixedPrecision(this.toJSON())
  }

  /**
   * Represents the state of your class as a JSON Object.
   *
   * @todo Implement this function in the derived class.
   */
  toJSON() {
    throw new Error(
      'Not yet implemented for this type:' + this.constructor.name
    )
  }
}

export { AttrValue }
export default AttrValue
