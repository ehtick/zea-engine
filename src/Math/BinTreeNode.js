import Vec2 from './Vec2.js';

class BinTreeRectBorder {
    constructor(value, movable = false) {
        this.value = value;
        this.movable = movable;
    }
};

class BinTreeRect {
    constructor(pos, right, top) {
        this.pos = pos;
        this.right = right;
        this.top = top;
    }

    get size() {
        return new Vec2(this.right.value - this.pos.x, this.top.value - this.pos.y);
    }
};

class BinTreeNode {
    constructor(rect, rootNode = false) {
        this.rect = rect;
        this.child0 = undefined;
        this.child1 = undefined;
        this.occupied = false;
        this.rootNode = rootNode;
        if (this.rootNode) {
            this.rect.right.ownerNode = this;
            this.rect.top.ownerNode = this;
        }
    }

    insert(rectSize, closestFit = undefined) {
        if (this.child0 != undefined && this.child1 != undefined) {
            // try inserting into first child
            let res = this.child0.insert(rectSize, closestFit);
            if (res != undefined)
                return res

            // no room, insert into second.
            return this.child1.insert(rectSize, closestFit);
        } else {

            // if there's already a rectSize here, return...
            if (this.occupied) {
                return undefined;
            }
            let delta = rectSize.subtract(this.rect.size);

            // (if we're too small, return)

            // Using a custom epsilon after packing was failing on simple cases. 
            // See TestBinTree2
            let eps = 0.00001;
            if (delta.x > eps || delta.y > eps) {
                // If we can fit the cluster into the node by moving only one of the sides.
                // Check for movable sides, and if so return this as a possibility of expanding the canvas.
                if (closestFit &&
                    ((this.rect.right.movable && delta.x > 0 && delta.y <= 0) ||
                     (this.rect.top.movable && delta.y > 0 && delta.x <= 0) ||
                     (this.rect.right.movable && delta.x > 0 && this.rect.top.movable && delta.y > 0 ))) {
                    
                    // Generate a cost metric to find the node that requires 
                    // the smallest ajustment of the borders.
                    let cost = 0;
                    if (this.rect.right.movable && delta.x > 0)
                        cost += delta.x;
                    if (this.rect.top.movable && delta.y > 0)
                        cost += delta.y;

                    if (!closestFit.node || cost < closestFit.cost) {
                        closestFit.node = this;
                        closestFit.delta = delta;
                        closestFit.cost = cost;
                    }
                }
                return undefined;
            }

            // if we're just right, accept..
            if (Math.abs(delta.x) < eps && Math.abs(delta.y) < eps) {
                if(this.rect.top.movable){
                    this.sliceHorizontal(this.rect.top.value);
                    this.child0.setOccupied();
                    return this.child0;
                }
                if(this.rect.right.movable){
                    this.sliceVertical(this.rect.right.value);
                    this.child0.setOccupied();
                    return this.child0;
                }
                this.setOccupied();
                return this;
            }

            // otherwise, gotta split this node and create some kids
            if (delta.x < delta.y)
                this.sliceVertical(this.rect.pos.x + rectSize.x);
            else
                this.sliceHorizontal(this.rect.pos.y + rectSize.y);

            return this.child0.insert(rectSize, closestFit);
        }
    }

    setOccupied() {
        this.occupied = true;
    }

    sliceVertical(xDivisor) {
        let child0rect = new BinTreeRect(this.rect.pos, new BinTreeRectBorder(xDivisor), this.rect.top);
        let child1rect = new BinTreeRect(new Vec2(xDivisor, this.rect.pos.y), this.rect.right, this.rect.top);
        this.child0 = new BinTreeNode(child0rect);
        this.child1 = new BinTreeNode(child1rect);
    }

    sliceHorizontal(yDivisor) {
        let child0rect = new BinTreeRect(this.rect.pos, this.rect.right, new BinTreeRectBorder(yDivisor));
        let child1rect = new BinTreeRect(new Vec2(this.rect.pos.x, yDivisor), this.rect.right, this.rect.top);
        this.child0 = new BinTreeNode(child0rect);
        this.child1 = new BinTreeNode(child1rect);
    }

    resizeAndInsert(rectSize, delta) {
        if (delta.x > 0 && this.rect.right.movable && delta.y > 0 && this.rect.top.movable) {
            this.rect.right.value += delta.x;
            this.rect.top.value += delta.y;
            this.sliceVertical(this.rect.right.value);
            this.child0.sliceHorizontal(this.rect.top.value);
            this.child0.child0.setOccupied();
            return this.child0.child0;
        }
        else if (delta.x > 0 && this.rect.right.movable){
            this.rect.right.value += delta.x;
            this.sliceVertical(this.rect.right.value);
            let res = this.child0.insert(rectSize);
            if (res == undefined)
                console.log("We should fit now~!!");
            return res;
        }
        else if (delta.y > 0 && this.rect.top.movable) {
            this.rect.top.value += delta.y;
            this.sliceHorizontal(this.rect.top.value);
            let res = this.child0.insert(rectSize);
            if (res == undefined)
                console.log("We should fit now~!!");
            return res;
        }
        else {
            throw ("Why are we here!!");
        }
    }
};

export {
    BinTreeRect,
    BinTreeRectBorder,
    BinTreeNode
};

export default BinTreeNode;
