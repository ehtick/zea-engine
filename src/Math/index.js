
export {
    SInt32,
    UInt32,
    Float32,
    hashStr,
    JSON_stringify_fixedPrecision
} from './Common.js';
export * from './AttrValue.js';
export * from './Vec2.js';
export * from './Vec3.js';
export * from './Vec4.js';
export * from './Color.js';
export * from './Quat.js';
export * from './Ray.js';
export * from './Mat3.js';
export * from './Mat4.js';
export * from './Xfo.js';
export * from './Box2.js';
export * from './Box3.js';
export * from './Rect2.js';
export * from './Hammersley.js';

export * from './TypeRegistry.js';
