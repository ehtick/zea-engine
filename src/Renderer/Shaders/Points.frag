
precision highp float;

#ifndef ENABLE_MULTI_DRAW

uniform color BaseColor;

#endif

import 'GLSLUtils.glsl'
import 'drawItemTexture.glsl'
import 'cutaways.glsl'
import 'materialparams.glsl'

#if defined(DRAW_GEOMDATA)

uniform int floatGeomBuffer;
uniform int passId;

import 'GLSLBits.glsl'

#elif defined(DRAW_HIGHLIGHT)

#ifdef ENABLE_FLOAT_TEXTURES
vec4 getHighlightColor(int id) {
  return fetchTexel(instancesTexture, instancesTextureSize, (id * pixelsPerItem) + 4);
}
#else // ENABLE_FLOAT_TEXTURES

uniform vec4 highlightColor;

vec4 getHighlightColor() {
  return highlightColor;
}

#endif // ENABLE_FLOAT_TEXTURES

#endif // DRAW_HIGHLIGHT

/* VS Outputs */
varying float v_drawItemId;
varying vec4 v_geomItemData;
varying vec3 v_viewPos;
/* VS Outputs */

#ifdef ENABLE_ES3
out vec4 fragColor;
#endif

void main(void) {

#ifndef ENABLE_ES3
  vec4 fragColor;
#endif

  //////////////////////////////////////////////
  // Color
#if defined(DRAW_COLOR)

#ifdef ENABLE_MULTI_DRAW

  vec2 materialCoords = v_geomItemData.zw;
  vec4 baseColor = getMaterialValue(materialCoords, 0);
  vec4 matValue1 = getMaterialValue(materialCoords, 1);
  float pointSize       = baseColor.a * matValue1.r;
  float overlay      = matValue1.g;

#else // ENABLE_MULTI_DRAW

  vec4 baseColor = BaseColor;

#endif // ENABLE_MULTI_DRAW

  fragColor = baseColor;

  //////////////////////////////////////////////
  // GeomData
#elif defined(DRAW_GEOMDATA)

  float viewDist = length(v_viewPos);

  if (floatGeomBuffer != 0) {
    fragColor.r = float(passId); 
    fragColor.g = float(v_drawItemId);
    fragColor.b = 0.0;// TODO: store poly-id or something.
    fragColor.a = viewDist;
  }
  else {
    ///////////////////////////////////
    // UInt8 buffer
    fragColor.r = mod(v_drawItemId, 256.) / 256.;
    fragColor.g = (floor(v_drawItemId / 256.) + (float(passId) * 64.)) / 256.;


    // encode the dist as a 16 bit float
    vec2 float16bits = encode16BitFloatInto2xUInt8(viewDist);
    fragColor.b = float16bits.x;
    fragColor.a = float16bits.y;
  }

  //////////////////////////////////////////////
  // Highlight
#elif defined(DRAW_HIGHLIGHT)
  
  int drawItemId = int(v_drawItemId + 0.5);
  fragColor = getHighlightColor(drawItemId);

#endif // DRAW_HIGHLIGHT


#ifndef ENABLE_ES3
  gl_FragColor = fragColor;
#endif
}
