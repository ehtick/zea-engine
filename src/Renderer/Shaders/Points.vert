
precision highp float;

attribute vec3 positions;

import 'GLSLUtils.glsl'
import 'geomItemId.glsl'
import 'drawItemTexture.glsl'
import 'modelMatrix.glsl'

uniform mat4 viewMatrix;
uniform mat4 projectionMatrix;

#ifdef ENABLE_MULTI_DRAW
import 'materialparams.glsl'
#else
uniform float PointSize;
uniform float Overlay;
#endif

/* VS Outputs */
varying float v_geomItemId;
varying vec4 v_geomItemData;
varying vec3 v_viewPos;

void main(void) {
  int geomItemId = getGeomItemId();
  v_geomItemId = float(geomItemId);
  v_geomItemData  = getInstanceData(geomItemId);

  mat4 modelMatrix = getModelMatrix(geomItemId);
  mat4 modelViewMatrix = viewMatrix * modelMatrix;
  
  vec4 viewPos = modelViewMatrix * vec4(positions, 1.);
  gl_Position = projectionMatrix * viewPos;
  

  //////////////////////////////////////////////
  // Material
#ifdef ENABLE_MULTI_DRAW
  vec2 materialCoords = v_geomItemData.zw;
  vec4 materialValue1 = getMaterialValue(materialCoords, 1);
  int maintainScreenSize = int(materialValue1.x + 0.5);
  float pointSize = materialValue1.x;
  float overlay = materialValue1.y;
#else
  float pointSize = PointSize;
  float overlay = Overlay;
#endif
  //////////////////////////////////////////////

  // Note: as of 22/01/2021 gl_PointSize has stopped working again...
  gl_PointSize = pointSize;

#if defined(DRAW_GEOMDATA)
  // Make the geom data point size at least 8 pixels across, else its impossible to hit.
  gl_PointSize = max(8.0, pointSize);
#endif
  gl_Position.z = mix(gl_Position.z, -gl_Position.w, overlay);

  
  v_viewPos = -viewPos.xyz;
}
