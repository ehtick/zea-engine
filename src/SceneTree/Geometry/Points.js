import { BaseGeom } from './BaseGeom.js';

class Points extends BaseGeom {
  constructor() {
    super();
  }

  loadBin(reader) {
    this.name = reader.loadStr();
    let numVerts = reader.loadUInt32();
    this.__boundingBox.set(reader.loadFloat32Vec3(), reader.loadFloat32Vec3());
    this.setNumVertices(numVerts);
    let vertices = this.vertices;

    if (numVerts < 256) {
      let bboxMat = this.__boundingBox.toMat4();
      let posAttr_8bit = reader.loadUInt8Array(numVerts*3);
      for (let i = 0; i<numVerts; i++) {
        let pos = new Vec3(
          posAttr_8bit[(i * 3) + 0] / 255.0, 
          posAttr_8bit[(i * 3) + 1] / 255.0, 
          posAttr_8bit[(i * 3) + 2] / 255.0
          );
        vertices.setValue(i, bboxMat.transformVec3(pos));
      }
    }
    else{
      let numClusters = reader.loadUInt32();
      let clusters = [];
      for (let i = 0; i < numClusters; i++) {
        let range = reader.loadUInt32Vec2();
        let p0 = reader.loadFloat32Vec3();
        let p1 = reader.loadFloat32Vec3();
        clusters.push({
          'range': range,
          'bbox': new Box3(p0, p1)
        });
      }
      let posAttr_8bit = reader.loadUInt8Array(numVerts*3);

      for (let i = 0; i < numClusters; i++) {
        let bboxMat = clusters[i]['bbox'].toMat4();
        for (let j = clusters[i]['range'].x; j < clusters[i]['range'].y; j++) {
          let pos = new Vec3(
            posAttr_8bit[(j * 3) + 0] / 255.0, 
            posAttr_8bit[(j * 3) + 1] / 255.0, 
            posAttr_8bit[(j * 3) + 2] / 255.0
            );
          vertices.setValue(j, bboxMat.transformVec3(pos));
        }
      }
    }
  }


  //////////////////////////////////////////
  // Persistence

  readBinary(reader, context) {
    super.loadBaseGeomBinary(reader);

    // this.computeVertexNormals();
    this.geomDataChanged.emit();
  }

};

export {
  Points
};
//export default Points;
