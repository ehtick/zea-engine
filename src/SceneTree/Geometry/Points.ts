import { BaseGeom } from './BaseGeom'
import { Registry } from '../../Registry'
import { BinReader } from '../BinReader'
import { Vec3 } from '../../Math/Vec3'
import { Box3 } from '../../Math/Box3'
import { Vec3Attribute } from './Vec3Attribute'
import { AssetLoadContext } from '../AssetLoadContext'

/**
 * Class representing a point primitive drawing type, every vertex specified is a point.
 *
 * ```
 * const points = new Points()
 * ```
 *
 * * **Events**
 * * **boundingBoxChanged:** Triggered when the bounding box changes.
 *
 * @extends BaseGeom
 */
class Points extends BaseGeom {
  /**
   * Create points.
   */
  constructor() {
    super()
  }

  /**
   * The clear method.
   */
  clear() {
    this.setNumVertices(0)
    this.emit('geomDataTopologyChanged')
  }

  // ////////////////////////////////////////
  // Persistence

  /**
   * Sets state of current geometry(Including line segments) using a binary reader object.
   *
   * @param reader - The reader value.
   * @param context - The context value.
   */
  readBinary(reader: BinReader, context: AssetLoadContext) {
    super.loadBaseGeomBinary(reader)

    // this.computeVertexNormals();
    this.emit('geomDataChanged')
  }
}

Registry.register('Points', Points)

export { Points }
