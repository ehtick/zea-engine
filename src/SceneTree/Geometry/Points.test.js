import { Points } from './Points'
import { Vec2, Vec3, Box3 } from '../../Math'

describe('Points', () => {
  test('Check for default positions attribute.', () => {
    const points = new Points()
    expect(points.hasVertexAttribute('positions')).toBe(true)
  })

  test('Check for positions attribute has the same length as the number of vertices.', () => {
    const points = new Points()
    const numVertices = 10
    points.setNumVertices(numVertices)
    expect(points.getVertexAttribute('positions').length).toBe(numVertices)
  })

  test('Check if new attribute has the same length as the number of vertices.', () => {
    const points = new Points()
    const numVertices = 10
    points.setNumVertices(numVertices)
    points.addVertexAttribute('foo', Vec2, 1.0)
    expect(points.getVertexAttribute('foo').length).toBe(numVertices)
    expect(points.getVertexAttribute('foo').getValueRef(0).toJSON()).toStrictEqual({ x: 1, y: 1 })
  })

  test('Check for calculation of bounding box.', () => {
    const points = new Points()
    const numVertices = 3
    points.setNumVertices(numVertices)
    const positions = points.getVertexAttribute('positions')
    positions.getValueRef(0).set(1, 2, 3)
    positions.getValueRef(1).set(-1, -2, -3)
    positions.getValueRef(2).set(2, 1, -3)

    points.setBoundingBoxDirty()

    const box3 = new Box3()
    box3.min.set(-1, -2, -3)
    box3.max.set(2, 2, 3)
    expect(points.getBoundingBox()).toEqual(box3)
  })

  test('Check resizing bigger the number of vertices.', () => {
    const points = new Points()

    points.setNumVertices(3)
    const positions = points.getVertexAttribute('positions')
    positions.getValueRef(0).set(1, 2, 3)
    positions.getValueRef(1).set(-1, -2, -3)
    positions.getValueRef(2).set(2, 1, -3)

    points.setNumVertices(4)
    expect(positions.getValueRef(0)).toEqual(new Vec3(1, 2, 3))
  })

  test('Check resizing smaller the number of vertices.', () => {
    const points = new Points()

    points.setNumVertices(3)
    const positions = points.getVertexAttribute('positions')
    positions.getValueRef(0).set(1, 2, 3)
    positions.getValueRef(1).set(-1, -2, -3)
    positions.getValueRef(2).set(2, 1, -3)

    points.setNumVertices(2)
    expect(positions.getValueRef(0)).toEqual(new Vec3(1, 2, 3))
  })

  test('Check generated buffers', () => {
    const points = new Points()
    const numVertices = 3
    points.setNumVertices(numVertices)
    const positions = points.getVertexAttribute('positions')
    positions.getValueRef(0).set(1, 2, 3)
    positions.getValueRef(1).set(-1, -2, -3)
    positions.getValueRef(2).set(2, 1, -3)

    expect(points.genBuffers()).toEqual({
      attrBuffers: {
        positions: {
          count: 3,
          dataType: Vec3,
          normalized: false,
          values: new Float32Array([1, 2, 3, -1, -2, -3, 2, 1, -3]),
        },
      },
      numVertices: 3,
    })
  })

  it('saves to JSON (serialization).', () => {
    const points = new Points()
    const numVertices = 3
    points.setNumVertices(numVertices)
    const positions = points.getVertexAttribute('positions')
    positions.getValueRef(0).set(1, 2, 3)
    positions.getValueRef(1).set(-1, -2, -3)
    positions.getValueRef(2).set(2, 1, -3)

    expect(JSON.stringify(points.toJSON())).toMatchSnapshot()
  })

  it('loads from JSON (serialization).', () => {
    const points = new Points()
    const input = {
      type: 'Points',
      numVertices: 3,
      vertexAttributes: {
        positions: { data: [1, 2, 3, -1, -2, -3, 2, 1, -3], dataType: 'Vec3', defaultValue: 0, length: 3 },
      },
    }
    points.fromJSON(input)

    expect(points.getVertexAttribute('positions').length).toBe(3)
  })
})
