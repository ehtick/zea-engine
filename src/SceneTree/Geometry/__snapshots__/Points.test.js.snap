// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Points saves to JSON (serialization). 1`] = `"{\\"type\\":\\"Points\\",\\"numVertices\\":3,\\"vertexAttributes\\":{\\"positions\\":{\\"data\\":[1,2,3,-1,-2,-3,2,1,-3],\\"dataType\\":\\"Vec3\\",\\"defaultValue\\":0,\\"length\\":3}}}"`;
